src/Microsoft.IdentityModel.Clients.ActiveDirectory/TokenCache.cs

```diff
@@ -407,32 +407,43 @@ internal AdalResultWrapper LoadFromCacheCommon(CacheQueryData cacheQueryData, Re
                     AdalTokenCacheKey cacheKey = kvp.Value.Key;
                     resultEx = kvp.Value.Value.Clone();
 
-                    bool tokenNearExpiry = (resultEx.Result.ExpiresOn <=
-                                            DateTime.UtcNow + TimeSpan.FromMinutes(ExpirationMarginInMinutes));
-                    bool tokenExtendedLifeTimeExpired = (resultEx.Result.ExtendedExpiresOn <=
-                                                         DateTime.UtcNow);
+                    bool tokenNearExpiry = resultEx.Result.ExpiresOn <=
+                                            DateTime.UtcNow + TimeSpan.FromMinutes(ExpirationMarginInMinutes);
+                    bool tokenExtendedLifeTimeExpired = resultEx.Result.ExtendedExpiresOn <= DateTime.UtcNow;
+                    bool tokenIsForSameResource = cacheKey.ResourceEquals(cacheQueryData.Resource);
+                    bool rtPresent = resultEx.RefreshToken != null;
+
+                    // Handle Broker case - RT is held by the broker and the request is for a different resource
+                    if (!tokenIsForSameResource && !rtPresent)
+                    {
+                        // Can't use the AT returned by the cache - it's for a different resource
+                        // and ADAL does not have the RT to fetch a new AT
```

Comment:
Is this true both for Android and iOS?

Reply:
Very good point, I'll try it out on iOS and get back to you.

Reply:
I can't test on iOS because I don't have a physical device and Apple does not allow to install apps on the emulators. I can work with Jenny or Travis to get this tested prior to the release.